garden_ai/notebook_metadata.py

```diff
@@ -221,6 +177,126 @@ def save_requirements_data(
+        disabled=False,
+    )
+
+    def base_image_observer(change):
+        with output:
+            nonlocal nb_meta
+            nb_meta.notebook_image_name = change.new.strip()
+            if nb_meta.notebook_image_name == "":
+                nb_meta.notebook_image_name = None
+            _pickle_metadata(nb_meta)
+
+    base_image_widget.observe(base_image_observer, "value")
+
+    # Requirements widget
+    req_widget = None
+    if nb_meta.notebook_requirements.file_format == "pip":
+        reqs_string = "\n".join([req for req in nb_meta.notebook_requirements.contents])
+        req_widget = widgets.Textarea(
+            value=reqs_string,
+            placeholder="Requirements",
+            layout=widgets.Layout(width="100%", height="80px"),
+            continuous_update=False,
+            disabled=False,
+        )
+    elif nb_meta.notebook_requirements.file_format == "conda":
+        req_widget = widgets.Textarea(
+            value=yaml.safe_dump(nb_meta.notebook_requirements.contents),
+            placeholder="Requirements",
+            layout=widgets.Layout(width="100%", height="300px"),
+            continuous_update=False,
+            disabled=False,
+        )
+
+    def reqs_observer(change):
+        with output:
+            nonlocal nb_meta
+            if nb_meta.notebook_requirements.file_format == "pip":
+                nb_meta.notebook_requirements.contents = change.new.split("\n")
+
+                if "" in nb_meta.notebook_requirements.contents:
+                    nb_meta.notebook_requirements.contents.remove("")
+                if nb_meta.notebook_requirements.contents == "":
+                    nb_meta.notebook_requirements.contents = None
+                _pickle_metadata(nb_meta)
+            if nb_meta.notebook_requirements.file_format == "conda":
+                nb_meta.notebook_requirements.contents = yaml.safe_load(change.new)
+                _pickle_metadata(nb_meta)
+
+    req_widget.observe(reqs_observer, "value")
+
+    # Requirements format widget
+    req_format_options = ["pip", "conda"]
+    req_format_widget = widgets.SelectMultiple(
+        options=req_format_options,
+        value=[nb_meta.notebook_requirements.file_format],
+        rows=len(req_format_options),
+        disabled=False,
+    )
+
+    def reqs_format_observer(change):
+        with output:
+            nonlocal nb_meta
+            nb_meta.notebook_requirements.file_format = change.new
+            _pickle_metadata(nb_meta)
+
+    req_format_widget.observe(reqs_format_observer, "value")
+
+    req_widget_tabs = widgets.Tab(
+        children=[req_widget, req_format_widget], titles=("Requirements", "Format")
+    )
+    metadata_widget = widgets.Accordion(
+        children=[doi_widget, base_image_widget, req_widget_tabs],
+        titles=("Global DOI", "Base Image", "Requirements"),
+    )
+
+    display(metadata_widget, output)
+
+
+def _pickle_metadata(nb_meta: NotebookMetadata):
```

Comment:
I really, really don't think we should be using pickle to persist anything we don't absolutely need to use pickle for. Because the data we're persisting here _can_ be JSON, it should be -- using pickle ties this metadata both to the python version used in the container and to the exact `NotebookMetadata` class definition we're using to read it, which is brittle. It would also be trivial to hand-build a valid .json for this but very hard to build a valid .pkl if any part of this breaks down.